### PR TITLE
Support refresh tokens

### LINKED ISSUE #12: Support refresh tokens
I'm curious if there's an interest to add support for refresh tokens to the oauth2 middleware. If there is I'd be happy to work on a PR to add this support.

Some OAuth2 providers will as part of the authorization flow return a refresh token. When an access token or idtoken expires the refresh token can be used to obtain new tokens. These new tokens might also contain a new refresh token. The following link describes the use of refresh tokens in more detail:

https://auth0.com/learn/refresh-tokens/

An application which would like to make use of the refresh token flow currently has a problem. It has access to the refresh token and could make a request for new tokens, but once this has been completed it doesn't have the possibility to write these newly retrieved tokens back to the session. The next request will not be able to use them.

The change I'd like to propose is that the middleware checks the expiry date of the accesstoken, and also of the idtoken if one exists. If either is expired one of two things happens:

- If a refresh token is available it is used to retrieve a new access token (and optionally id token). These replace the old tokens in the session.
- If no refresh token is available the session gets invalidated and a new authorization flow is started.

I'd love to hear your thoughts on this proposal!

### PR DESCRIPTION
**Pull request by jwoudenberg**
Resolves #12

This adds supports for refresh tokens to the OAuth2 middleware.

It also splits out a dedicated OpenID Connect middleware, because it's session invalidation logic is slightly different. Up to this point I've been using the oauth2 middleware for an OpenID Connect flow, making use of the fact that we can fetch an ID token by sending an `idtoken` scope and then read that id token out of the session. To add refresh token support we need to be able to figure out when a session has expired though, and here the OAuth2 and OpenID implementations diverge:

- For the OAuth2 flow we check the expire time of the access token.
- For the OpenID Connect flow we check the expire time of the id token.

I think existing users of the OAuth2 provider probably wouldn't want their sessions to become invalidated because of some id token they might not use or know about. Conversely those looking to rely on ID tokens might not need an access token or be interested in its expiry.

I also added some test for old and new functionality of the OAuth2 and OpenID Connect providers. Happy to look into setting up some CI (maybe based on Travis?) if that seems useful!

Please let me know what you think!

### CODE REVIEW

**Comment by snoyberg** [comment]
You interested in being a comaintainer of this package?

**Comment by jwoudenberg** [comment]
Sure, happy to help! Thank you for offering!

**Comment by snoyberg** [comment]
Invite sent here on Github. What's your Hackage username?

**Comment by jwoudenberg** [comment]
It's `JasperWoudenberg`

**Comment by snoyberg** [comment]
@lehins can you grant @jwoudenberg (and myself if convenient) maintainer access on Hackage?

**Comment by lehins** [comment]
Done. You are both maintainers of this package on hackage.

**Comment by jwoudenberg** [comment]
Quick update on my radio silence here: I've been wanting to try this out on a local project before calling it 'done'. I've recently started doing that and found a couple of small issues. Working on fixing those, and then I'll get back here!

**Comment by jwoudenberg** [comment]
This seems to work 🎉. @lehins would you be interested in reviewing? I realize that as a maintainer I'm able to move this forward myself, but I'm not sure about the PR etiquette in this repo!

File: src/Network/Wai/Middleware/Auth/OIDC.hs
Comment (lehins) [inline src/Network/Wai/Middleware/Auth/OIDC.hs]: I think it is ok to drop the support for jose-0.7.x right away in favor of reduced complexity.
Moreover, jose-0.8.0 has been out for a year and a half and is already in 3 stackage lts snapshots.

File: src/Network/Wai/Middleware/Auth/OIDC.hs
Comment (lehins) [inline src/Network/Wai/Middleware/Auth/OIDC.hs]: One of uses for the auth providers is ability to read their configuration from a yaml file. Sticking a manager into the config will make it impossible to give it a FromJSON instance. See Google+OAuth2 for example: 
https://github.com/fpco/wai-middleware-auth/blob/c9544ab9ad631c612f7940340c3d9c56a0c219dc/src/Network/Wai/Middleware/Auth/OAuth2/Google.hs#L72-L84

File: src/Network/Wai/Middleware/Auth/OIDC.hs
Comment (lehins) [inline src/Network/Wai/Middleware/Auth/OIDC.hs:102]: We need to be able to grab the config from a yaml file, which is required for `wai-auth` binary. If we are to add OpenID Connect, we need a `FromJSON` instance for this data type.

File: src/Network/Wai/Middleware/Auth/OAuth2.hs
Comment (lehins) [inline src/Network/Wai/Middleware/Auth/OAuth2.hs:125]: It is better to move the lines above here since neither `oauth2` nor `man` are needed in the `else` case:

```haskell
    authEndpointURI <- parseAbsoluteURI oa2AuthorizeEndpoint
    accessTokenEndpointURI <- parseAbsoluteURI oa2AccessTokenEndpoint
    let oauth2 =
          OA2.OAuth2
          { oauthClientId = getClientId oa2ClientId
          , oauthClientSecret = getClientSecret oa2ClientSecret
          , oauthOAuthorizeEndpoint = authEndpointURI
          , oauthAccessTokenEndpoint = accessTokenEndpointURI
          -- Setting callback endpoint to `Nothing` below is a lie.
          -- We do have a callback endpoint but in this context
          -- don't have access to the function that can render it.
          -- We get away with this because the callback endpoint is
          -- not needed for obtaining a refresh token, the only
          -- way we use the config here constructed.
          , oauthCallback = Nothing
          }
    man <- getGlobalManager
```

File: src/Network/Wai/Middleware/Auth/OAuth2.hs
Comment (lehins) [inline src/Network/Wai/Middleware/Auth/OAuth2.hs]: Feel free to leave it as is if you like, but this is just a `fmap`, so I think it be cleaner with it, something along those lines:

```haskell
              pure (rRes <&> \ newTokens -> (req, user {
                      authLoginState = encodeToken newTokens,
                      authLoginTime = fromIntegral now
                    }))
```

**Review by lehins — APPROVED** [review]
@jwoudenberg Sorry it took a while to respond to your PR. 

There are a couple minor improvements I suggested here, but overall looks real good. Once you take care of those suggestions feel free to rebase on master, wait for CI to pass and then release it on Hackage as 0.2.3.0.

I noticed that there was never a 0.2.2.0 release on hackage, I guess we can just skip it.

**Comment by jwoudenberg** [comment]
Hi @lehins, thank you so much for your feedback, and sorry for taking so long to make use of it! Going through it now. Will make updates and then release as you suggested!

File: src/Network/Wai/Middleware/Auth/OIDC.hs
Comment (jwoudenberg) [reply src/Network/Wai/Middleware/Auth/OIDC.hs]: Makes sense!

File: src/Network/Wai/Middleware/Auth/OIDC.hs
Comment (jwoudenberg) [reply src/Network/Wai/Middleware/Auth/OIDC.hs:102]: Cool, will do this!

File: src/Network/Wai/Middleware/Auth/OIDC.hs
Comment (jwoudenberg) [reply src/Network/Wai/Middleware/Auth/OIDC.hs]: That makes sense! Reason I put the `Manager` in there is that I'd like to be able to use a customized version of the manager in our application (one that adds tracing to outgoing requests).

I think both use cases can be served by letting the `FromJSON` instance always insert the global manager. Going to give that a try!

File: src/Network/Wai/Middleware/Auth/OAuth2.hs
Comment (jwoudenberg) [reply src/Network/Wai/Middleware/Auth/OAuth2.hs:125]: Makes sense, thanks!

File: src/Network/Wai/Middleware/Auth/OAuth2.hs
Comment (jwoudenberg) [reply src/Network/Wai/Middleware/Auth/OAuth2.hs]: Yep, better. Thank you!

File: src/Network/Wai/Middleware/Auth/OIDC.hs
Comment (jwoudenberg) [reply src/Network/Wai/Middleware/Auth/OIDC.hs:102]: Good news! Looks like Aeson has bundled a `FromJSON NominalDiffTime` instance for a couple of years, so there's nothing for us to do here!

File: src/Network/Wai/Middleware/Auth/OIDC.hs
Comment (jwoudenberg) [reply src/Network/Wai/Middleware/Auth/OIDC.hs]: Fixed by making `oidcManager` a `Maybe Manager`.

**Comment by jwoudenberg** [comment]
Addressed feedback, rebased on master, and CI is green! Going to merge, then release!

**Comment by jwoudenberg** [comment]
Released as 0.2.3.0

**Comment by lehins** [comment]
@jwoudenberg Great job! Thank you for the new release ;)

**Pull request merged**
